Upgrade to Electron v0.30.6

thomasjo opened the pull request:
This is a continuation of #6877.
## Blocking issues
- [x] Building with io.js v2.x is not supported.
- [x] Crash caused by invalid memory access during exit/closing window.
- [x] 8 failing spec assertions.
- [x] In OS X closing all windows exits the application.
- [x] Building on Linux (Ubuntu 12.04.5 LTS) is failing due to compilation errors.
- [x] Failing settings-view specs.
- [x] apm builds incompatible native modules.
- [x] Experiment with upgrading to Electron v0.31.x.  
  **Conclusion**: Not viable in this PR since it requires bumping every single dependency to NAN 2.x.
- [x] Wait until https://github.com/atom/incompatible-packages/pull/9 is ready to merge.
- [x] Preserve `CustomEvent` as type of object dispatched by command registry
- [x] Investigate why symbols-view is failing to find any symbols
- [x] Investigate extremely long paint times for cursor movement on first-gen MBP retinas
### Issue details
#### Building with io.js v2.x is not supported

**Resolution**: Do not update npm. The version shipped with io.js is patched to work around the problems with node-gyp.

Any attempt to execute `script/build` when using io.js v2.x fails with

```
Installing build modules...
gyp ERR! configure error
gyp ERR! stack Error: 404 response downloading https://nodejs.org/dist/v2.4.0/node-v2.4.0.tar.gz
gyp ERR! stack     at Request.<anonymous> (/Users/thomasjo/.nvm/versions/io.js/v2.4.0/lib/node_modules/npm/node_modules/node-gyp/lib/install.js:251:14)
gyp ERR! stack     at emitOne (events.js:82:20)
gyp ERR! stack     at Request.emit (events.js:169:7)
gyp ERR! stack     at Request.onRequestResponse (/Users/thomasjo/.nvm/versions/io.js/v2.4.0/lib/node_modules/npm/node_modules/request/request.js:985:10)
gyp ERR! stack     at emitOne (events.js:77:13)
gyp ERR! stack     at ClientRequest.emit (events.js:169:7)
gyp ERR! stack     at HTTPParser.parserOnIncomingClient (_http_client.js:415:21)
gyp ERR! stack     at HTTPParser.parserOnHeadersComplete (_http_common.js:88:23)
gyp ERR! stack     at TLSSocket.socketOnData (_http_client.js:305:20)
gyp ERR! stack     at emitOne (events.js:77:13)
gyp ERR! System Darwin 14.4.0
gyp ERR! command "/Users/thomasjo/.nvm/versions/io.js/v2.4.0/bin/iojs" "/Users/thomasjo/.nvm/versions/io.js/v2.4.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/thomasjo/Code/atom/atom-electron/build/node_modules/runas
gyp ERR! node -v v2.4.0
gyp ERR! node-gyp -v v2.0.2
gyp ERR! not ok
```
#### Crash caused by invalid memory access during exit/closing window

**Resolution**: Solved in Electron v0.30.2.
[Full crash log](https://gist.github.com/thomasjo/8080e027a291c9402292).
The problem is a bit tricky to reproduce, but when it occurs it happens when a window is closed or when exiting Atom, which leads me to believe the problem is related to storing state. Seems like the crash is caused by a `nullptr` reference.
#### 8 failing spec assertions

[Full output](https://gist.github.com/thomasjo/541278db1f5ac7675a87) from `apm test`. Summary of failing specs;
##### command-registry-spec.coffee
- [x] `Expected undefined to be HTMLNode` at command-registry-spec.coffee:212:16.  
  _Caused by https://code.google.com/p/chromium/issues/detail?id=495437._
##### package-manager-spec.coffee

All failures originate from `it triggers the activation event on all handlers registered during activation`.
- [x] `Expected 0 to be 1`.
- [x] `Expected 0 to be 1`.
- [x] `Expected 0 to be 1`.
- [x] `Expected 0 to be 2`.
- [x] `Expected 0 to be 2`.
- [x] `Expected 0 to be 1`.
##### pane-container-view-spec.coffee
- [x] `Expected [ undefined ] to equal [ { 0 : HTMLNode, element : <getter>, activeItemModifiedChanged : Function, ...` at pane-container-view-spec.coffee:60:21.
#### In OS X closing all windows exits the application

**Resolution**: Side-effect of crashing problem. Resolved in Electron v0.30.2.

When running against atom-shell v0.22.3 Atom behaves as expected and leaves the application running even after all application windows have been closed. When running against Electron v0.30.0 however, after all windows have been closed the application exits.
#### Building on Linux (Ubuntu 12.04.5 LTS) is failing due to compilation errors

**Resolution**: Using GCC 4.7+, or Clang 3.0+. Also fixed in CI by ensuring the usage of Clang instead of GCC.

Attempts to build on Ubuntu 12.04.5 LTS (64-bit) fails while compiling npm modules, throwing errors similar to

```
gyp ERR! cwd /home/travis/build/atom/atom/node_modules/oniguruma
gyp ERR! node -v v0.10.35
gyp ERR! node-gyp -v v1.0.2
gyp ERR! not ok 
In file included from /home/travis/.atom/.node-gyp/.node-gyp/0.30.1/src/node.h:42:0,
                 from ../node_modules/nan/nan.h:24,
                 from ../src/repository.h:29,
                 from ../src/repository.cc:22:
/home/travis/.atom/.node-gyp/.node-gyp/0.30.1/deps/v8/include/v8.h:543:1: error: expected unqualified-id before ‘using’
/home/travis/.atom/.node-gyp/.node-gyp/0.30.1/deps/v8/include/v8.h:922:1: error: expected unqualified-id before ‘using’
```

The compilation failure is caused by the fact that GCC 4.7 or newer is required due to V8 using the _template alias_ feature in C++, and that is not supported in GCC 4.6.3 which is the newest GCC (officially) available in Ubuntu 12.04.5 LTS.

**Solution**: Compile with Clang 3.4 on Ubuntu 12.04.5 LTS;

```
$ sudo apt-get install clang-3.4
$ CXX=clang++ script/build
```

For Travis CI support, we should specify `compiler: clang` in `.travis.yml`. (http://docs.travis-ci.com/user/languages/c/#Choosing-compilers-to-test-against)
#### Failing settings-view specs

Resolution: Fixed bugs in settings-view package@0.211.3.

```
ThemesPanel
  when a UI theme is selected
    it updates the 'core.themes' config key with the selected UI theme
      Expected [ 'atom-dark-syntax' ] to equal [ 'atom-light-ui', 'atom-dark-syntax' ]. (spec/themes-panel-spec.coffee:53:46)
  when a syntax theme is selected
    it updates the 'core.themes' config key with the selected syntax theme
      Expected [ 'atom-dark-ui' ] to equal [ 'atom-dark-ui', 'atom-light-syntax' ]. (spec/themes-panel-spec.coffee:58:46)
  when the 'core.config' key changes
    it refreshes the theme menus
      Expected 'atom-dark-ui' to be 'atom-light-ui'. (spec/themes-panel-spec.coffee:69:36)
      Expected 'atom-dark-syntax' to be 'atom-light-syntax'. (spec/themes-panel-spec.coffee:70:40)


Finished in 17.811 seconds
101 tests, 440 assertions, 4 failures, 0 skipped
```
#### apm builds incompatible native modules.

**Resolution**: Solved by apm v1.0.3.

When apm builds native modules they end up incompatible with Atom. See https://github.com/atom/atom/pull/7877#issuecomment-127031350 for more details.
## Other potential problems
1. ~~Currently it does not seem like it's possible to use this build of Atom as a Git editor; closing the spawned window does not result in the process exiting as expected it seems like, and thus leaves the Git terminal process waiting for the input sequence to end.  
   This is possibly related to issue _In OS X closing all windows exits the application_.~~ Resolved by upgrading to Electron v0.30.2.

---

Fixes https://github.com/atom/atom/issues/6335
Fixes https://github.com/atom/atom/issues/3826
Fixes https://github.com/atom/atom/issues/7502
Fixes https://github.com/atom/atom/issues/3636
Fixes https://github.com/atom/atom/issues/6243
Fixes https://github.com/atom/atom/issues/6306
Fixes https://github.com/atom/atom/issues/6219
Fixes https://github.com/atom/atom/issues/7811
Fixes https://github.com/atom/atom/issues/7374
Fixes https://github.com/atom/atom/issues/8309
Fixes https://github.com/atom/atom/issues/6513

kevinsawicki committed:
Use grunt-download-electron

kevinsawicki committed:
Update build to use electron

kevinsawicki committed:
Add ~/.atom/electron to clean script

kevinsawicki committed:
atomShellVersion -> electronVersion

kevinsawicki committed:
Ignore root electron folder

kevinsawicki committed:
Rename atom shell tasks

kevinsawicki committed:
Remove log suppressing

kevinsawicki committed:
Copy electron folder

kevinsawicki committed:
Rename Electron exe to Atom

kevinsawicki committed:
Uncoment install task

kevinsawicki committed:
:arrow_up: onigiruma@4.2.2

kevinsawicki committed:
:arrow_up: grunt-cson@0.15

kevinsawicki committed:
Re-enable install

kevinsawicki committed:
Rename Electron Helper to Atom Helper

kevinsawicki committed:
Add --no-install option to script/build

kevinsawicki committed:
atom-shell -> electron

kevinsawicki committed:
Rename electron executable on Linux and Windows

kevinsawicki committed:
Set ATOM_NODE_VERSION env var from electronVersion in package.json

kevinsawicki committed:
Add a ;

kevinsawicki committed:
:arrow_up: electron@0.26.1

zcbenz committed:
:arrow_up: asar@0.7.0

zcbenz committed:
:arrow_up: electron@0.27.2

zcbenz committed:
Object.create doesn't work well with CustomEvent in Chrome 43

thomasjo committed:
:arrow_up: electron@0.29.2

jlord commented:
Woot! I ran the branch on my ARM Chromebook, got the same errors you mention, but it did everything before then which is great, had trouble building #6877.

lee-dohm commented:
I've been using this branch lightly this morning. Haven't had any issues so far.

```
ares: "1.10.1-DEV"
atom-shell: "0.29.2"
chrome: "43.0.2357.65"
electron: "0.29.2"
http_parser: "2.5.0"
modules: "44"
node: "2.3.1"
openssl: "1.0.2c"
uv: "1.6.1"
v8: "4.3.61.21"
zlib: "1.2.8"
```

lee-dohm commented:
Started running into an issue. Here's what I did. Haven't tried to generalize or minimize the repro yet, but wanted to get this out there.

## Repro Steps

1. `git revert HEAD` &rarr; launches a new window
1. Save with <kbd>Cmd+S</kbd>
1. Close window with <kbd>Cmd+Shift+W</kbd>

**Expected:** Atom closes the window but does not exit
**Actual:** Atom closes the window and then crashes

No matter how I close the window, Atom crashes.

Crash logs are here: https://gist.github.com/lee-dohm/12798a98827934084766

thomasjo commented:
> No matter how I close the window, Atom crashes.
> 
> Crash logs are here: https://gist.github.com/lee-dohm/12798a98827934084766

Yeah that's the same type of crash I've been getting all along. I've tried safe mode which doesn't help. Might try soft-nuking the entire `.atom` folder to see if it's related to persisted state. It definitely looks like a memory leak or an invalid pointer since we're seeing

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
```

I haven't been able to track down the exact issue yet. Maybe @zcbenz has any idea based on the crash log?

zcbenz commented:
@thomasjo You can revert 2818cd1 and9f7b8cf since they were a try to make V8 snapshot work in Atom but didn't work well. I'll look into the crash.

thomasjo commented:
@zcbenz Thanks, I rebased the branch and yanked out the snapshot commits.

The crash problem seems much harder to reproduce after the rebase. Whether coincidental or not, is unknown to me at this point.

thomasjo committed:
:arrow_up: electron@0.30.0

thomasjo commented:
I'm having problems using Atom as a "Git Editor" using this branch; quitting the Atom instance spawned via e.g. `git commit` seems to just leave Git waiting for the process to exit. Can anyone else reproduce that behavior?

thomasjo committed:
:arrow_up: electron@0.30.1

thomasjo committed:
:see_no_evil: Add temporary hack to work around issue with CustomEvent

Since it's no longer possible to "properly" extend CustomEvent, we will need to
figure out a way to shadow the prototype chain, or some such.

See https://code.google.com/p/chromium/issues/detail?id=495437 for more details.

thomasjo commented:
The (temporary) hack I added in 1982167315ac15944443562a5837539df3389bf8, fixes the failing core specs. There are still failing package specs though (`archive-view` and `settings-view`).

thomasjo commented:
@zcbenz Have you had a chance to investigate the crash? Also, has anything changed in Electron on OS X since v0.22.3 with regards to closing last window also exiting the application? Thanks :bow:

thomasjo commented:
Compiling with Clang 3.4 seems to have resolved the compilation issues on Ubuntu 12.04.5 LTS.

zcbenz commented:
For the not building with io.js v2.x issue, I think you need to update your npm, otherwise it can not find the headers of iojs at the right place.

zcbenz commented:
> @zcbenz Have you had a chance to investigate the crash?

I have fixed the crash on master branch, will publish a new release and update here soon.

> Also, has anything changed in Electron on OS X since v0.22.3 with regards to closing last window also exiting the application? Thanks

I think it is because Atom crashed when closing a window.

thomasjo commented:
@zcbenz Awesome! Thanks for your help :bow:

zcbenz commented:
> I'm having problems using Atom as a "Git Editor" using this branch; quitting the Atom instance spawned via e.g. git commit seems to just leave Git waiting for the process to exit. Can anyone else reproduce that behavior?

This is also caused by the crash when closing window, after fixing the crash I don't see this anymore.

zcbenz committed:
Install gcc 4.8 on travis CI

zcbenz committed:
Use clang to build native modules

paulcbetts commented:
Getting close!! :clap::clap::clap::clap::clap::clap::clap:

zcbenz committed:
Use clang as compiler on travis CI

thomasjo commented:
> For the not building with io.js v2.x issue, I think you need to update your npm, otherwise it can not find the headers of iojs at the right place.

But I'm using the latest npm (v2.13.2) and that has no effect. If I use io.js v1.8.4, it works just fine. What combination of io.js and npm are you using that works?

thomasjo commented:
@zcbenz Turns out the problem is with `node-gyp`. There are two open PRs with a fix:
 1. https://github.com/TooTallNate/node-gyp/pull/564
 2. https://github.com/TooTallNate/node-gyp/pull/669

zcbenz commented:
We now only have a few settings-view specs failing in CI.

thomasjo commented:
> We now only have a few settings-view specs failing in CI.

:sparkles: Thanks for the assistance @zcbenz! I'll look at the settings-view specs now.

thomasjo commented:
@zcbenz For the problems with node-gyp and io.js, I found a reasonable workaround that I'm using locally; I downloaded the tarball and used the `--tarball` option to point to that instead of letting `node-gyp` download it (and fail). Do you reckon that's a reasonable workaround for the time being? I was planning to automate this in `script/bootstrap` and use it whenever the dev is using io.js (at least for `>=2.0.0`).

thomasjo commented:
Looking at the settings-view with this build, there is something really odd going on with the drop-drop lists on the themes panel; selecting new themes seems to work, but the selection doesn't "stick". A bit tricky to explain, but very easy to reproduce.

zcbenz committed:
:arrow_up: electron@0.30.2

zcbenz commented:
> Do you reckon that's a reasonable workaround for the time being? I was planning to automate this in script/bootstrap and use it whenever the dev is using io.js (at least for >=2.0.0).

You can make use of `npm_config_disturl` to change where to download headers:

```
export npm_config_disturl=https://iojs.org/dist
```

thomasjo commented:
> You can make use of npm_config_disturl to change where to download headers:

I tried that, and it doesn't work for two reasons; 
 1. This line in node-gyp: https://github.com/TooTallNate/node-gyp/blob/master/lib/install.js#L188
 2. Setting the `dist-url` option like that also affects apm's npm which is currently targeting Node.js v0.10.35, so the headers are completely wrong in that case.

Until this gets properly resolved in node-gyp, I think the most reasonable approach is to handle the downloading ourselves, or fork and patch node-gyp.

zcbenz commented:
It seems that you have to use the npm that shipped in io.js, I thought upstream npm already supported io.js but it turns out I was wrong.

thomasjo commented:
> It seems that you have to use the npm that shipped in io.js, I thought upstream npm already supported io.js but it turns out I was wrong.

Aha, I thought so too. Mentions of patched npm seemed to have dissipated, so I thought it was merged. Hopefully soon!

kevinsawicki commented on script/cibuild:
```diff
@@ -22,11 +22,16 @@ function loadEnvironmentVariables(filePath) {
     loadEnvironmentVariables('/var/lib/jenkins/config/xcodekeychain');
+  } else if (process.platform === 'linux') {
+    // Use clang for building native code, the GCC on precise is too old.
+    process.env['CC'] = 'clang';
```
Super minor, but this could be rewritten as `process.env.CC` to be consistent with other env property access in this file.

kevinsawicki commented on src/command-registry.coffee:
```diff
@@ -175,11 +175,14 @@ class CommandRegistry
       stopImmediatePropagation: value: ->
+    # In Chrome 43, Object.create doesn't work well with CustomEvent.
+    for k, v of event when k not in eventWithTarget
+      eventWithTarget[k] = v
```
Could this use `?=` instead?

```coffee
eventWithTarget[key] ?= value for key, value of event
```

Or even:

```coffee
_.defaults(eventWithTarget, event)
```

thomasjo replied:
This whole block needs to change anyway, since all properties have been moved into the prototypes, which is why added e3174cd. I'll need to shadow the prototype properties from `CustomEvent` (only one IIRC) and `Event`.

thomasjo committed:
:art: Set environment variables using dot notation

anaisbetts committed:
Set App User Model ID on Win32

thomasjo commented:
Believe I've fixed the settings-view bugs: https://github.com/atom/settings-view/pull/616. Just need to get it reviewed and :ship:. Once that's done I'll rebase on top of master once more and resolve the conflicts.

mnquintana commented:
Just ran into a regression - `apm` seems to have some issues properly rebuilding modules with native extensions.

To reproduce:

1. Clone bracket-matcher if you haven't already
1. rm -rf bracket-matcher/node_modules
1. `apm install`
1. `apm link -d`
1. Open Atom, and you'll get this error for bracket-matcher:
```
oniguruma@4.2.2
Error message: Module version mismatch. Expected 44, got 43.
```

thomasjo commented:
@mnquintana Thanks for the report. I'm looking into this. Not entirely sure, but we might have to upgrade apm's Node?

zcbenz commented:
Yeah we also need to bump the ATOM_NODE_VERSION in apm.

mnquintana commented:
Thanks @zcbenz!

Just opened https://github.com/atom/apm/issues/408 which does seem to resolve the last issue with building native modules.

lee-dohm commented:
I just got this error message when launching Atom from the Dock (not sure it has anything to do with it, but figured I'd include it) and now it is happening any time I try to launch Atom. Atom launches, shows this error ... and that's it. No window ... but it doesn't exit either:

<img width="532" alt="screen shot 2015-08-03 at 8 11 21 am" src="https://cloud.githubusercontent.com/assets/1038121/9040562/dd14c5e8-39b7-11e5-9bac-f05783079883.png">

thomasjo commented:
@lee-dohm Yeah that's the native module problem. Thanks anyway!

thomasjo committed:
Simplify workaround, and describe issue with Object.create + CustomEvent

thomasjo committed:
:arrow_up: electron@0.30.3

thomasjo committed:
:arrow_up: electron@0.30.4

YuriSolovyov commented:
one tiny nit: latest 0.30.x electron is [0.30.5](https://github.com/atom/electron/releases/tag/v0.30.5)

thomasjo committed:
:arrow_up: electron@0.30.5

thomasjo commented:
> one tiny nit: latest 0.30.x electron is 0.30.5

It was released about 10 hours ago, so give me a little slack :wink: 
Thanks for paying attention!

thomasjo commented:
@zcbenz Do you have a feeling for whether the [bump to Chrome 44](https://github.com/atom/electron/pull/2474) will cause a lot of problems on this end? I'm a bit reluctant to bump Electron to v0.31.0, but I'll give it a try locally and see what fails.

thomasjo commented:
When bumping to Electron v0.31.0, the Nan Issue™ seems quite prevalent and prevents building. Bumping runas to v3.1.0 in `build/package.json` makes it possible to build apm. However, the fact that git-utils is using nan v1.6.x prevents it from being built. There might be other packages that are currently not building, but since the build stops on failure I have no easy means of finding the other failures.

mnquintana commented:
IMO we should just keep this at 0.30.x so we can :ship: soon, and worry about the next upgrade in another PR. :smile:

zcbenz commented:
Chrome 44 will break all of our native modules, we have to update every module to use NAN v2 to fix this. And it is unfortunate that NAN v2 changed its API significantly so this progress would be quite painful.

thomasjo commented:
> Chrome 44 will break all of our native modules, we have to update every module to use NAN v2 to fix this. And it is unfortunate that NAN v2 changed its API significantly so this progress would be quite painful.

OK, that seals the deal then. I'm going to kill the v0.31.x experiment in this PR. We should get this one shipped soon, then we can begin working on a new v0.31.x PR.

kevinsawicki commented on package.json:
```diff
@@ -41,8 +41,8 @@
     "nslog": "^2.0.0",
-    "oniguruma": "^4.1",
     "pathwatcher": "^4.4.3",
+    "oniguruma": "^4.2.2",
```
Hmm, this looks to have moved to be out of alphabetical order now

kevinsawicki commented on script/bootstrap:
```diff
@@ -83,6 +83,9 @@ function bootstrap() {
     apmInstallCommand += ' --arch=ia32';
 
+  if (!process.env.ATOM_NODE_VERSION)
+    process.env.ATOM_NODE_VERSION = require('../package.json').electronVersion;
```
This shouldn't be needed anymore right since `apm` reads `electronVersion`?

If we still want it as a fallback, it should probably be renamed to the new `ATOM_ELECTRON_VERSION` env var support in apm.

kevinsawicki commented:
Just a heads up, this PR will be blocked until https://github.com/atom/incompatible-packages/pull/9 is merged which allows people to rebuild their modules from within Atom to make the upgrade easier.

thomasjo committed:
:art: Fix non-alphabetical ordering of dependencies

thomasjo committed:
:fire: Remove unnecessary ATOM_NODE_VERSION bootstrapping

capouch commented:
Electron version has bumped again, this time to 0.31.0.

YuriSolovyov commented:
@capouch yes, but it brings a lot of breakage for native modules since they all need to be updated to NaN 2, please read the thread.

capouch commented:
LOL I read it ten times a day my entire workgroup is stalled on deploying Atom until it is fixed, because the current version will not run under a VM environment.  It was a reminder that we need to change the name of those issue yet again.

hastebrot commented:
@capouch 

>It was a reminder that we need to change the name of this issue yet again.

Nope. 0.30.5 is still the latest version of Electron 0.30.x.

capouch commented:
@hastebrot 
** snip **
$ electron --version
[1528:0822/141901:ERROR:browser_main_loop.cc(185)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
v0.31.0

thomasjo commented:
> Or are you saying they won't move past 0.30.x at present?

Correct. See https://github.com/atom/atom/pull/7877#issuecomment-133456156.

hastebrot commented:
>Or are you saying they won't move past 0.30.x at present?

Yes. The two comments before your first comment tell you everything.

Commit:
Continue to dispatch CustomEvent instances in CommandRegistry

This changes the approach used to synthesize an event to dispatch. Since
we can no longer point to the original event as our prototype, we build
a new CustomEvent instance and override select properties with
Object.defineProperty. This preserves the contract with event handlers
by giving them a true CustomEvent instance.

nathansobo commented:
@kevinsawicki If you've tested out the process names, this should be ready to merge after the next release. I addressed the dispatch issue from `CommandRegistry` like we discussed.

nathansobo commented:
I'd like to merge this immediate following the merge of https://github.com/atom/atom/pull/7877. Then we can upgrade the `incompatible-packages` package on that branch and test everything for a while on master.

thomasjo committed:
:bug: Fix undefined reference to syntheticEvent

thomasjo commented:
Broken symbols-view is caused by a regression in ASAR via https://github.com/atom/asar/commit/771a7be9063a4dba14925ec5b172e5b44f366170. There's a [proposed PR](https://github.com/atom/asar/pull/41) that should resolve the problem.

nathansobo commented:
:zap: @thomasjo :zap:

nathansobo commented:
Okay, the horrific paint times seemed to have been solved by doing an [SMC reset](https://support.apple.com/en-us/HT201295) on my machine. Once I did that, I started looking really critically at paint times and *maybe* perceived things being slightly faster with `--num-raster-threads=1`, but it's barely perceptible so I say we roll forward. What a day. :finnadie:

paulcbetts commented:
> Okay, the horrific paint times seemed to have been solved by doing an SMC reset on my machine. Once I did that, I started looking really critically at paint times and maybe perceived things being slightly faster with --num-raster-threads=1, but it's barely perceptible so I say we roll forward. What a day

Not sure I believe you, have you seen https://github.com/atom/electron/issues/2332?

nathansobo commented:
@paulcbetts It might be *slightly* degraded in a way I'm having a tough time measuring, but it's noting like what I was seeing before the reset. What I was seeing would have been a disaster.

paulcbetts commented:
@nathansobo We should still seriously consider putting out an 0.30.6 release with that Brightray bug fixed or else Atom paint perf will go through the floor /cc @zcbenz

nathansobo commented:
> We should still seriously consider putting out an 0.30.6 release with that Brightray bug fixed or else Atom paint perf will go through the floor

:+1:

zcbenz commented:
@nathansobo Can you try downgrading Electron to 0.30.0 and see if the performance issues still exists? I want to verify it is caused by the brightray bug, if it is I will publish a 0.30.6 release to fix that.

paulcbetts commented:
Scrolling fast will show you the difference, framerate will go way down

nathansobo commented:
@zcbenz After downgrading to 0.30.0, I can confirm what @paulcbetts is saying. The cursor movements slowness seemed to be due to an issue with my machine, but for fast scrolling in a large file, I'm seeing a pretty substantial difference. This isn't the most controlled test as I'm just scrolling with my trackpad manually, but here's a comparison scrolling through jQuery.

#### 0.30.5

![electron-0 30 5-fps](https://cloud.githubusercontent.com/assets/1789/9484835/c1e0fe18-4b6e-11e5-8c37-f9b16a7c8e89.gif)


#### 0.30.0

![electron-0 30 0-fps](https://cloud.githubusercontent.com/assets/1789/9484836/cc08d334-4b6e-11e5-895c-8c2f4ea323c8.gif)

thomasjo committed:
:arrow_up: electron@0.30.6

zackarychapple commented:
Bump, when can this get merged?

v3ss0n commented:
Need to wait from main chromium's fix i guess.

thomasjo commented:
> Bump, when can this get merged?

Basically when the core team is confident that potential regressions in the latest version of Atom have been resolved. Since it's entirely possible that merging this might cause lots of issues, it's important to ensure that timing is right.

> Need to wait from main chromium's fix i guess.

Nope, nothing like that. See above.

nathansobo commented:
If we don't hear about any regressions over the weekend, I'm comfortable merging Monday. Sorry for the delay. We're working on a better release system so we don't have to block master like this going forward.

v3ss0n commented:
@nathansobo  and atom team ! theres no need to be sorry , you guys doing awesome work and i understand , i can wait.

Zireael07 commented:
Seconding @v3ss0n. This is awesome!

sLite commented:
So looking forward to this being merged :)

High five from all virtualbox users!

hak8or commented:
Echoing what @sLite said, I eagerly await this merge so I can finally use Atom in a virtual machine without it thinking I am always using a touchscreen device, as described [here](https://discuss.atom.io/t/how-do-i-stop-atom-from-displaying-a-touchscreen-insertion-point-when-run-in-a-vm-under-a-host-that-is-not-touch-enabled/19942) and issue #6306.

nathansobo commented:
Thanks everyone for your patience. Onward and upward. :chart_with_upwards_trend:

wzxu commented:
Kudos guys! Can't wait for the next version to land so I can use Atom in VM again. Wonder if there's any performance gain as well upgrading Chrome?

paulcbetts commented:
:clap: :clap: :clap: :clap: :clap: :clap: :clap: :clap: :clap: :clap:

despairblue commented:
:fireworks: finally I change my default browser back to chrome :bow:

capouch commented:
@despairblue is it working all right for you?  It does not work in my VirtualBox instance.  The previous versions were workable but lacked click-drag-select.  My initial testing with this one does not show any difference; I still can't do it.

sLite commented:
I built directly from this branch today before it got merged, worked like a charm on Arch Linux in a VirtualBox.

zackarychapple commented:
Fantastic!! Good job!

batjko commented:
woot :fireworks:

v3ss0n commented:
:clap: :clapper:  :clapper: :clapper: :clapper: :clapper: :clapper: :clapper: :clapper:

v3ss0n commented:
Bump Atom to 1.1! worth it!

hastebrot commented:
@v3ss0n Hmm, does the update of Electron cause backwards-incompatible changes that justify a minor version bump?

mehcode commented:
> @v3ss0n Hmm, does the update of Electron cause backwards-incompatible changes that justify a minor version bump?

Because this just broke any native packages still on NAN 1.x.

despairblue commented:
@capouch I was actually referring to [this](https://github.com/atom/electron/issues/1348#issuecomment-91230583), which would lock up atom on linux when clicking a link if your default browser is chrome or chromium. Electron fixed this in April and this PR pulls these changes in.

I don't use atom in a VM so I can't say anything about this.

capouch commented:
@despairblue it built just fine for me, but actually runs worse than the previous version.  I am running it on Slackware 64-bit VirtualBox 4.3.28, Electron 0.32.0 is also installed on the machine so maybe that is causing the issues, but the mouse sprites are all over the place, and click-drag unchanged from earlier.  Built this afternoon from the github master.

marciobda commented:
Awesome work! It built just fine for me too. 
I´m running it on Ubuntu 14.04 VirtualBox 5.0.2

capouch commented:
@marciobda have you tested that you can click-drag to select text in the VM environment?

marciobda commented:
Yes. I was watching this thread specifically because that click-drag problem in VMs.

v3ss0n commented:
As far as i Understand , this will break any packages using NaN 1.x right?  So yes 1.1 is appropriate i think.

wzxu commented:
@v3ss0n I think only Electron 0.31.0+ breaks NaN 1.x, which is why they settled at 0.30.6 for now. See https://github.com/atom/atom/pull/7877#issuecomment-133456156.

v3ss0n commented:
Ah i c.

martgnz commented:
I just updated to 1.0.11 and I still have the click-drag problem. Ubuntu 15.04 on VirtualBox.

jacobnbrum commented:
1.0.11 is still the old version.

martgnz commented:
When is this going to be released?

jacobnbrum commented:
I am not sure but you can try building it yourself: https://github.com/atom/atom/blob/master/docs/build-instructions/linux.md

capouch commented:
@marciobda what is your precise operating environment?  It shouldn't matter that I have Electron 0.32.0 and a API v46 Chrome installed on the same VM?  I'm running 64bit Slackware ("current" as of mid-August) on VirtualBox 4.3.28.  It's actually worse now--the mouse sprite doesn't change properly as I navigate.  Built from sources; the build itself went perfectly.

hak8or commented:
I recommend making a new github issue or continuing issue #6306 for further discussion about this. While I used to have the same issue (I didn't get a chance to test from master yet), I am pretty sure the dev's who worked on things relevant to this specific pull request would prefer to keep things more organized.
